File: src/linter-integration/tool-management-system.ts
/**
 * Tool Management System
 * Manages linter tool lifecycle, configuration, and resource allocation
 * MESH NODE AGENT: Integration Specialist for Linter Integration Architecture Swarm
 */

import { EventEmitter } from 'events';
import { spawn, ChildProcess } from 'child_process';
import { promises as fs } from 'fs';
import { join, resolve } from 'path';
import { performance } from 'perf_hooks';

// Import types from the main engine
import { LinterTool, CircuitBreakerState } from './real-time-ingestion-engine';

interface ToolEnvironment {
  nodeVersion?: string;
  pythonVersion?: string;
  environmentVariables: Record<string, string>;
  workingDirectory: string;
  pathExtensions: string[];
}

interface ToolConfiguration {
  configFile?: string;
  rules?: Record<string, any>;
  ignore?: string[];
  include?: string[];
  customArgs?: string[];
  environment?: ToolEnvironment;
}

interface ResourceAllocation {
  cpuLimit?: number;
  memoryLimit?: number;
  concurrencyLimit: number;
  priorityWeight: number;
  executionQuota: number;
  throttleInterval: number;
}

interface ToolHealth {
  isHealthy: boolean;
  lastHealthCheck: number;
  healthScore: number;
  failureRate: number;
  averageExecutionTime: number;
  successfulExecutions: number;
  failedExecutions: number;
  lastError?: string;
}

interface ToolMetrics {
  totalExecutions: number;
  successfulExecutions: number;
  failedExecutions: number;
  averageExecutionTime: number;
  minExecutionTime: number;
  maxExecutionTime: number;
  totalViolationsFound: number;
  uniqueRulesTriggered: Set<string>;
  resourceUsage: {
    peakMemory: number;
    totalCpuTime: number;
    diskUsage: number;
  };
}

interface ToolRecoveryProcedure {
  resetConfiguration: boolean;
  reinstallTool: boolean;
  clearCache: boolean;
  restartEnvironment: boolean;
  escalateToAdmin: boolean;
  customRecoverySteps: string[];
}

/**
 * Tool Management System
 * Comprehensive lifecycle management for linter tools
 */
export class ToolManagementSystem extends EventEmitter {
  private readonly tools: Map<string, LinterTool> = new Map();
  private readonly configurations: Map<string, ToolConfiguration> = new Map();
  private readonly environments: Map<string, ToolEnvironment> = new Map();
  private readonly resourceAllocations: Map<string, ResourceAllocation> = new Map();
  private readonly healthStatus: Map<string, ToolHealth> = new Map();
  private readonly metrics: Map<string, ToolMetrics> = new Map();
  private readonly circuitBreakers: Map<string, CircuitBreakerState> = new Map();
  private readonly recoveryProcedures: Map<string, ToolRecoveryProcedure> = new Map();
  
  private readonly runningProcesses: Map<string, ChildProcess> = new Map();
  private readonly executionQueue: Map<string, Array<() => Promise<void>>> = new Map();
  
  private readonly maxGlobalConcurrency: number = 10;
  private readonly healthCheckInterval: number = 30000;
  private readonly metricsRetentionPeriod: number = 86400000; // 24 hours
  
  constructor(private readonly workspaceRoot: string) {
    super();
    this.initializeDefaultEnvironments();
    this.initializeDefaultResourceAllocations();
    this.setupPeriodicTasks();
  }

  /**
   * Initialize default tool environments
   */
  private initializeDefaultEnvironments(): void {
    // Node.js environment for TypeScript/JavaScript tools
    this.environments.set('nodejs', {
      nodeVersion: '18.0.0',
      environmentVariables: {
        NODE_ENV: 'production',
        NODE_OPTIONS: '--max-old-space-size=4096'
      },
      workingDirectory: this.workspaceRoot,
      pathExtensions: ['node_modules/.bin']
    });

    // Python environment for Python tools
    this.environments.set('python', {
      pythonVersion: '3.8.0',
      environmentVariables: {
        PYTHONPATH: this.workspaceRoot,
        PYTHONUNBUFFERED: '1',
        PYTHONDONTWRITEBYTECODE: '1'
      },
      workingDirectory: this.workspaceRoot,
      pathExtensions: ['.venv/bin', 'venv/Scripts']
    });

    // Generic system environment
    this.environments.set('system', {
      environmentVariables: {},
      workingDirectory: this.workspaceRoot,
      pathExtensions: []
    });
  }

  /**
   * Initialize default resource allocations
   */
  private initializeDefaultResourceAllocations(): void {
    const defaultAllocations: Record<string, ResourceAllocation> = {
      eslint: {
        concurrencyLimit: 3,
        priorityWeight: 0.8,
        executionQuota: 100,
        throttleInterval: 1000
      },
      tsc: {
        concurrencyLimit: 1, // TypeScript compiler is resource-intensive
        priorityWeight: 0.9,
        executionQuota: 50,
        throttleInterval: 2000
      },
      flake8: {
        concurrencyLimit: 2,
        priorityWeight: 0.7,
        executionQuota: 80,
        throttleInterval: 1500
      },
      pylint: {
        concurrencyLimit: 1, // Pylint is slow
        priorityWeight: 0.6,
        executionQuota: 30,
        throttleInterval: 3000
      },
      ruff: {
        concurrencyLimit: 4, // Ruff is very fast
        priorityWeight: 0.8,
        executionQuota: 150,
        throttleInterval: 500
      },
      mypy: {
        concurrencyLimit: 2,
        priorityWeight: 0.7,
        executionQuota: 60,
        throttleInterval: 2000
      },
      bandit: {
        concurrencyLimit: 2,
        priorityWeight: 0.9, // Security is high priority
        executionQuota: 70,
        throttleInterval: 1500
      }
    };

    Object.entries(defaultAllocations).forEach(([toolId, allocation]) => {
      this.resourceAllocations.set(toolId, allocation);
    });
  }

  /**
   * Setup periodic health checks and maintenance tasks
   */
  private setupPeriodicTasks(): void {
    // Health check interval
    setInterval(() => {
      this.performHealthChecks();
    }, this.healthCheckInterval);

    // Metrics cleanup interval
    setInterval(() => {
      this.cleanupOldMetrics();
    }, this.metricsRetentionPeriod / 24); // Check every hour

    // Resource usage monitoring
    setInterval(() => {
      this.monitorResourceUsage();
    }, 5000); // Every 5 seconds
  }

  /**
   * Register a new linter tool
   */
  public async registerTool(
    tool: LinterTool, 
    configuration?: ToolConfiguration
  ): Promise<void> {
    try {
      // Validate tool installation
      await this.validateToolInstallation(tool);
      
      // Store tool and configuration
      this.tools.set(tool.id, tool);
      if (configuration) {
        this.configurations.set(tool.id, configuration);
      }
      
      // Initialize health status and metrics
      this.initializeToolHealth(tool.id);
      this.initializeToolMetrics(tool.id);
      
      // Setup circuit breaker
      this.circuitBreakers.set(tool.id, {
        isOpen: false,
        failureCount: 0,
        lastFailureTime: 0,
        successCount: 0,
        nextAttemptTime: 0
      });
      
      // Setup recovery procedures
      this.setupRecoveryProcedures(tool);
      
      this.emit('tool_registered', { toolId: tool.id, name: tool.name });
      
    } catch (error) {
      this.emit('tool_registration_failed', { 
        toolId: tool.id, 
        error: error.message 
      });
      throw error;
    }
  }

  /**
   * Validate that a tool is properly installed and accessible
   */
  private async validateToolInstallation(tool: LinterTool): Promise<void> {
    return new Promise((resolve, reject) => {
      const testCommand = tool.healthCheckCommand || `${tool.command} --version`;
      const [command, ...args] = testCommand.split(' ');
      
      const process = spawn(command, args, {
        stdio: ['pipe', 'pipe', 'pipe'],
        timeout: 10000
      });
      
      let stdout = '';
      let stderr = '';
      
      process.stdout.on('data', (data) => {
        stdout += data.toString();
      });
      
      process.stderr.on('data', (data) => {
        stderr += data.toString();
      });
      
      process.on('close', (code) => {
        if (code === 0) {
          resolve();
        } else {
          reject(new Error(`Tool validation failed for ${tool.name}: ${stderr || stdout}`));
        }
      });
      
      process.on('error', (error) => {
        reject(new Error(`Failed to execute ${tool.name}: ${error.message}`));
      });
    });
  }

  /**
   * Initialize health status for a tool
   */
  private initializeToolHealth(toolId: string): void {
    this.healthStatus.set(toolId, {
      isHealthy: true,
      lastHealthCheck: Date.now(),
      healthScore: 100,
      failureRate: 0,
      averageExecutionTime: 0,
      successfulExecutions: 0,
      failedExecutions: 0
    });
  }

  /**
   * Initialize metrics for a tool
   */
  private initializeToolMetrics(toolId: string): void {
    this.metrics.set(toolId, {
      totalExecutions: 0,
      successfulExecutions: 0,
      failedExecutions: 0,
      averageExecutionTime: 0,
      minExecutionTime: Infinity,
      maxExecutionTime: 0,
      totalViolationsFound: 0,
      uniqueRulesTriggered: new Set(),
      resourceUsage: {
        peakMemory: 0,
        totalCpuTime: 0,
        diskUsage: 0
      }
    });
  }

  /**
   * Setup recovery procedures for a tool
   */
  private setupRecoveryProcedures(tool: LinterTool): void {
    const procedures: ToolRecoveryProcedure = {
      resetConfiguration: true,
      reinstallTool: false,
      clearCache: true,
      restartEnvironment: false,
      escalateToAdmin: false,
      customRecoverySteps: []
    };

    // Tool-specific recovery procedures
    switch (tool.id) {
      case 'eslint':
        procedures.customRecoverySteps = [
          'npm install eslint --save-dev',
          'rm -rf node_modules/.cache/eslint'
        ];
        break;
      case 'tsc':
        procedures.customRecoverySteps = [
          'npm install typescript --save-dev',
          'rm -rf node_modules/.cache/tsc'
        ];
        break;
      case 'flake8':
      case 'pylint':
      case 'mypy':
      case 'bandit':
        procedures.customRecoverySteps = [
          `pip install --upgrade ${tool.id}`,
          'rm -rf __pycache__',
          'rm -rf .mypy_cache'
        ];
        break;
    }

    this.recoveryProcedures.set(tool.id, procedures);
  }

  /**
   * Execute a tool with full lifecycle management
   */
  public async executeTool(
    toolId: string, 
    filePaths: string[], 
    options: ToolExecutionOptions = {}
  ): Promise<ToolExecutionResult> {
    const tool = this.tools.get(toolId);
    if (!tool) {
      throw new Error(`Tool not found: ${toolId}`);
    }

    const allocation = this.resourceAllocations.get(toolId)!;
    const health = this.healthStatus.get(toolId)!;
    const circuitBreaker = this.circuitBreakers.get(toolId)!;

    // Check circuit breaker
    if (circuitBreaker.isOpen) {
      if (Date.now() < circuitBreaker.nextAttemptTime) {
        throw new Error(`Circuit breaker open for tool ${toolId}`);
      }
    }

    // Check health status
    if (!health.isHealthy && !options.forceExecution) {
      throw new Error(`Tool ${toolId} is unhealthy. Use forceExecution to override.`);
    }

    // Wait for resource availability
    await this.waitForResourceAvailability(toolId);

    const startTime = performance.now();
    let executionResult: ToolExecutionResult;

    try {
      // Execute with resource monitoring
      executionResult = await this.executeWithMonitoring(tool, filePaths, options);
      
      // Update success metrics
      this.updateSuccessMetrics(toolId, performance.now() - startTime, executionResult);
      
      // Reset circuit breaker on success
      circuitBreaker.failureCount = 0;
      circuitBreaker.successCount++;
      circuitBreaker.isOpen = false;
      
      return executionResult;
      
    } catch (error) {
      // Update failure metrics
      this.updateFailureMetrics(toolId, performance.now() - startTime, error);
      
      // Update circuit breaker
      circuitBreaker.failureCount++;
      circuitBreaker.lastFailureTime = Date.now();
      
      if (circuitBreaker.failureCount >= 5) {
        circuitBreaker.isOpen = true;
        circuitBreaker.nextAttemptTime = Date.now() + 60000; // 1 minute
        
        // Attempt recovery
        await this.attemptToolRecovery(toolId);
      }
      
      throw error;
    }
  }

  /**
   * Wait for resource availability based on allocation limits
   */
  private async waitForResourceAvailability(toolId: string): Promise<void> {
    const allocation = this.resourceAllocations.get(toolId)!;
    const runningCount = this.getRunningProcessCount(toolId);
    
    if (runningCount >= allocation.concurrencyLimit) {
      return new Promise((resolve) => {
        const queue = this.executionQueue.get(toolId) || [];
        queue.push(resolve);
        this.executionQueue.set(toolId, queue);
      });
    }
  }

  /**
   * Execute tool with comprehensive monitoring
   */
  private async executeWithMonitoring(
    tool: LinterTool, 
    filePaths: string[], 
    options: ToolExecutionOptions
  ): Promise<ToolExecutionResult> {
    const environment = this.getToolEnvironment(tool);
    const configuration = this.configurations.get(tool.id);
    
    // Prepare execution arguments
    const args = this.prepareExecutionArgs(tool, filePaths, configuration, options);
    
    return new Promise((resolve, reject) => {
      const startTime = performance.now();
      const startMemory = process.memoryUsage();
      
      const childProcess = spawn(tool.command, args, {
        env: { ...process.env, ...environment.environmentVariables },
        cwd: environment.workingDirectory,
        stdio: ['pipe', 'pipe', 'pipe']
      });
      
      this.runningProcesses.set(`${tool.id}_${Date.now()}`, childProcess);
      
      let stdout = '';
      let stderr = '';
      
      childProcess.stdout.on('data', (data) => {
        stdout += data.toString();
      });
      
      childProcess.stderr.on('data', (data) => {
        stderr += data.toString();
      });
      
      childProcess.on('close', (code) => {
        const executionTime = performance.now() - startTime;
        const endMemory = process.memoryUsage();
        const memoryUsed = endMemory.heapUsed - startMemory.heapUsed;
        
        if (code === 0 || (code === 1 && stdout)) {
          resolve({
            success: true,
            output: stdout,
            stderr,
            executionTime,
            memoryUsed,
            exitCode: code,
            violationsFound: this.countViolationsInOutput(tool, stdout)
          });
        } else {
          reject(new Error(`Tool execution failed with code ${code}: ${stderr}`));
        }
        
        // Clean up process reference
        this.runningProcesses.delete(`${tool.id}_${Date.now()}`);
        
        // Process next in queue
        this.processExecutionQueue(tool.id);
      });
      
      childProcess.on('error', (error) => {
        reject(error);
        this.runningProcesses.delete(`${tool.id}_${Date.now()}`);
        this.processExecutionQueue(tool.id);
      });
      
      // Set timeout
      setTimeout(() => {
        if (!childProcess.killed) {
          childProcess.kill('SIGTERM');
          reject(new Error(`Tool execution timed out after ${tool.timeout}ms`));
        }
      }, tool.timeout);
    });
  }

  /**
   * Perform health checks on all registered tools
   */
  private async performHealthChecks(): Promise<void> {
    const healthPromises = Array.from(this.tools.keys()).map(toolId => 
      this.performToolHealthCheck(toolId)
    );
    
    await Promise.allSettled(healthPromises);
  }

  /**
   * Perform health check on a specific tool
   */
  private async performToolHealthCheck(toolId: string): Promise<void> {
    const tool = this.tools.get(toolId)!;
    const health = this.healthStatus.get(toolId)!;
    
    try {
      await this.validateToolInstallation(tool);
      
      health.isHealthy = true;
      health.lastHealthCheck = Date.now();
      health.healthScore = Math.min(100, health.healthScore + 10);
      
      this.emit('tool_health_ok', { toolId, healthScore: health.healthScore });
      
    } catch (error) {
      health.isHealthy = false;
      health.lastHealthCheck = Date.now();
      health.healthScore = Math.max(0, health.healthScore - 20);
      health.lastError = error.message;
      
      this.emit('tool_health_degraded', { 
        toolId, 
        error: error.message, 
        healthScore: health.healthScore 
      });
      
      // Attempt recovery if health is critically low
      if (health.healthScore <= 20) {
        await this.attemptToolRecovery(toolId);
      }
    }
  }

  /**
   * Attempt to recover a failed tool
   */
  private async attemptToolRecovery(toolId: string): Promise<void> {
    const procedures = this.recoveryProcedures.get(toolId);
    if (!procedures) return;
    
    this.emit('tool_recovery_started', { toolId });
    
    try {
      if (procedures.resetConfiguration) {
        await this.resetToolConfiguration(toolId);
      }
      
      if (procedures.clearCache) {
        await this.clearToolCache(toolId);
      }
      
      // Execute custom recovery steps
      for (const step of procedures.customRecoverySteps) {
        await this.executeRecoveryStep(step);
      }
      
      // Reinitialize tool health
      this.initializeToolHealth(toolId);
      
      this.emit('tool_recovery_completed', { toolId });
      
    } catch (error) {
      this.emit('tool_recovery_failed', { toolId, error: error.message });
      
      if (procedures.escalateToAdmin) {
        this.emit('tool_recovery_escalation', { toolId, error: error.message });
      }
    }
  }

  /**
   * Get comprehensive tool status
   */
  public getToolStatus(toolId: string): ToolStatus {
    const tool = this.tools.get(toolId);
    const health = this.healthStatus.get(toolId);
    const metrics = this.metrics.get(toolId);
    const circuitBreaker = this.circuitBreakers.get(toolId);
    const allocation = this.resourceAllocations.get(toolId);
    
    if (!tool || !health || !metrics || !circuitBreaker || !allocation) {
      throw new Error(`Tool not found: ${toolId}`);
    }
    
    return {
      tool,
      health,
      metrics,
      circuitBreaker,
      allocation,
      isRunning: this.getRunningProcessCount(toolId) > 0,
      queueLength: (this.executionQueue.get(toolId) || []).length
    };
  }

  /**
   * Get status of all tools
   */
  public getAllToolStatus(): Record<string, ToolStatus> {
    const status: Record<string, ToolStatus> = {};
    
    this.tools.forEach((_, toolId) => {
      status[toolId] = this.getToolStatus(toolId);
    });
    
    return status;
  }

  // Helper methods
  private getToolEnvironment(tool: LinterTool): ToolEnvironment {
    // Determine environment based on tool type
    if (['eslint', 'tsc'].includes(tool.id)) {
      return this.environments.get('nodejs')!;
    } else if (['flake8', 'pylint', 'mypy', 'bandit', 'ruff'].includes(tool.id)) {
      return this.environments.get('python')!;
    } else {
      return this.environments.get('system')!;
    }
  }

  private prepareExecutionArgs(
    tool: LinterTool, 
    filePaths: string[], 
    configuration?: ToolConfiguration, 
    options?: ToolExecutionOptions
  ): string[] {
    let args = [...tool.args];
    
    if (configuration?.customArgs) {
      args = [...args, ...configuration.customArgs];
    }
    
    if (options?.additionalArgs) {
      args = [...args, ...options.additionalArgs];
    }
    
    args = [...args, ...filePaths];
    
    return args;
  }

  private getRunningProcessCount(toolId: string): number {
    return Array.from(this.runningProcesses.keys())
      .filter(key => key.startsWith(`${toolId}_`))
      .length;
  }

  private processExecutionQueue(toolId: string): void {
    const queue = this.executionQueue.get(toolId) || [];
    if (queue.length > 0) {
      const nextExecution = queue.shift()!;
      nextExecution();
      this.executionQueue.set(toolId, queue);
    }
  }

  private countViolationsInOutput(tool: LinterTool, output: string): number {
    // Basic violation counting logic - would be enhanced per tool
    try {
      if (tool.outputFormat === 'json') {
        const data = JSON.parse(output);
        if (Array.isArray(data)) {
          return data.reduce((count, file) => count + (file.messages?.length || 0), 0);
        }
      }
    } catch (error) {
      // Fallback to line counting for text output
      return output.split('\n').filter(line => line.trim()).length;
    }
    return 0;
  }

  private updateSuccessMetrics(toolId: string, executionTime: number, result: ToolExecutionResult): void {
    const metrics = this.metrics.get(toolId)!;
    const health = this.healthStatus.get(toolId)!;
    
    metrics.totalExecutions++;
    metrics.successfulExecutions++;
    metrics.totalViolationsFound += result.violationsFound;
    
    // Update execution time statistics
    const currentAvg = metrics.averageExecutionTime;
    metrics.averageExecutionTime = (currentAvg * (metrics.totalExecutions - 1) + executionTime) / metrics.totalExecutions;
    metrics.minExecutionTime = Math.min(metrics.minExecutionTime, executionTime);
    metrics.maxExecutionTime = Math.max(metrics.maxExecutionTime, executionTime);
    
    // Update health
    health.successfulExecutions++;
    health.averageExecutionTime = metrics.averageExecutionTime;
    health.failureRate = metrics.failedExecutions / metrics.totalExecutions;
  }

  private updateFailureMetrics(toolId: string, executionTime: number, error: Error): void {
    const metrics = this.metrics.get(toolId)!;
    const health = this.healthStatus.get(toolId)!;
    
    metrics.totalExecutions++;
    metrics.failedExecutions++;
    
    health.failedExecutions++;
    health.failureRate = metrics.failedExecutions / metrics.totalExecutions;
    health.lastError = error.message;
  }

  private async resetToolConfiguration(toolId: string): Promise<void> {
    // Implementation for resetting tool configuration
  }

  private async clearToolCache(toolId: string): Promise<void> {
    // Implementation for clearing tool cache
  }

  private async executeRecoveryStep(step: string): Promise<void> {
    // Implementation for executing recovery commands
  }

  private cleanupOldMetrics(): void {
    // Implementation for cleaning up old metrics data
  }

  private monitorResourceUsage(): void {
    // Implementation for monitoring resource usage
  }
}

// Additional interfaces
interface ToolExecutionOptions {
  forceExecution?: boolean;
  additionalArgs?: string[];
  timeout?: number;
  priority?: 'low' | 'medium' | 'high' | 'critical';
}

interface ToolExecutionResult {
  success: boolean;
  output: string;
  stderr: string;
  executionTime: number;
  memoryUsed: number;
  exitCode: number;
  violationsFound: number;
}

interface ToolStatus {
  tool: LinterTool;
  health: ToolHealth;
  metrics: ToolMetrics;
  circuitBreaker: CircuitBreakerState;
  allocation: ResourceAllocation;
  isRunning: boolean;
  queueLength: number;
}

export {
  ToolManagementSystem,
  ToolConfiguration,
  ToolEnvironment,
  ResourceAllocation,
  ToolHealth,
  ToolMetrics,
  ToolExecutionOptions,
  ToolExecutionResult,
  ToolStatus
};

File: src/linter_integration/tool_management_system.py
"""Tool lifecycle management primitives for linter integration tests."""

from __future__ import annotations

import asyncio
import json
import time
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Dict, List, Optional


@dataclass
class ToolEnvironment:
    nodeVersion: Optional[str] = None
    pythonVersion: Optional[str] = None
    environmentVariables: Dict[str, str] = field(default_factory=dict)
    workingDirectory: str = ""
    pathExtensions: List[str] = field(default_factory=list)


@dataclass
class ToolConfiguration:
    configFile: Optional[str] = None
    rules: Dict[str, Any] = field(default_factory=dict)
    ignore: List[str] = field(default_factory=list)
    include: List[str] = field(default_factory=list)
    customArgs: List[str] = field(default_factory=list)
    environment: Optional[ToolEnvironment] = None


@dataclass
class ResourceAllocation:
    cpuLimit: Optional[float] = None
    memoryLimit: Optional[int] = None
    concurrencyLimit: int = 1
    priorityWeight: float = 0.5
    executionQuota: int = 50
    throttleInterval: int = 1000


@dataclass
class ToolHealth:
    isHealthy: bool = True
    healthScore: int = 100
    failureRate: float = 0.0
    averageExecutionTime: float = 0.0
    successfulExecutions: int = 0
    failedExecutions: int = 0
    lastError: Optional[str] = None


@dataclass
class ToolMetrics:
    totalExecutions: int = 0
    successfulExecutions: int = 0
    failedExecutions: int = 0
    averageExecutionTime: float = 0.0
    minExecutionTime: float = float("inf")
    maxExecutionTime: float = 0.0
    totalViolationsFound: int = 0
    uniqueRulesTriggered: set[str] = field(default_factory=set)


@dataclass
class ToolExecutionOptions:
    additionalArgs: List[str] = field(default_factory=list)


@dataclass
class ToolExecutionResult:
    success: bool
    output: str
    stderr: str
    executionTime: float
    memoryUsed: int
    exitCode: int
    violationsFound: int


@dataclass
class RecoveryProcedures:
    resetConfiguration: bool = True
    clearCache: bool = True
    customRecoverySteps: List[str] = field(
        default_factory=lambda: ["restart_process", "refresh_environment"]
    )


@dataclass
class CircuitBreaker:
    isOpen: bool = False
    failureCount: int = 0
    lastFailureTime: float = 0.0
    successCount: int = 0
    nextAttemptTime: float = 0.0


@dataclass
class ToolStatus:
    tool: Any
    health: ToolHealth
    metrics: ToolMetrics
    circuitBreaker: Any
    allocation: ResourceAllocation
    isRunning: bool
    queueLength: int


class ToolManagementSystem:
    def __init__(self, workspaceRoot: str) -> None:
        self.workspaceRoot = workspaceRoot
        self.maxGlobalConcurrency = 10
        self.environments = self._default_environments(workspaceRoot)
        self.resourceAllocations = self._default_allocations()
        self.tools: Dict[str, Any] = {}
        self.configurations: Dict[str, ToolConfiguration] = {}
        self.healthStatus: Dict[str, ToolHealth] = {}
        self.metrics: Dict[str, ToolMetrics] = {}
        self.circuitBreakers: Dict[str, Any] = {}
        self.recoveryProcedures: Dict[str, RecoveryProcedures] = {}
        self.runningProcesses: Dict[str, Any] = {}
        self.executionQueue: Dict[str, List[Any]] = {}
        self._semaphores: Dict[str, asyncio.Semaphore] = {}

    def _default_environments(self, workspace: str) -> Dict[str, ToolEnvironment]:
        return {
            "nodejs": ToolEnvironment(
                nodeVersion="18",
                workingDirectory=workspace,
                environmentVariables={"NODE_ENV": "test"},
                pathExtensions=["node_modules/.bin"],
            ),
            "python": ToolEnvironment(
                pythonVersion="3",
                workingDirectory=workspace,
                environmentVariables={"PYTHONPATH": workspace},
                pathExtensions=[".venv/bin", ".venv/Scripts"],
            ),
            "system": ToolEnvironment(workingDirectory=workspace),
        }

    def _default_allocations(self) -> Dict[str, ResourceAllocation]:
        return {
            "eslint": ResourceAllocation(concurrencyLimit=2, priorityWeight=0.7),
            "tsc": ResourceAllocation(concurrencyLimit=1, priorityWeight=0.8),
            "flake8": ResourceAllocation(
                concurrencyLimit=2,
                priorityWeight=0.7,
                executionQuota=80,
                throttleInterval=1500,
            ),
            "pylint": ResourceAllocation(
                concurrencyLimit=1,
                priorityWeight=0.6,
                executionQuota=30,
                throttleInterval=3000,
            ),
            "ruff": ResourceAllocation(
                concurrencyLimit=4,
                priorityWeight=0.9,
                executionQuota=150,
                throttleInterval=500,
            ),
            "mypy": ResourceAllocation(concurrencyLimit=2, priorityWeight=0.8),
            "bandit": ResourceAllocation(concurrencyLimit=2, priorityWeight=0.8),
        }

    async def registerTool(
        self, tool: Any, config: Optional[ToolConfiguration] = None
    ) -> None:
        await self.validateToolInstallation(tool)
        self.tools[tool.id] = tool
        self.configurations[tool.id] = config or ToolConfiguration()
        self.initializeToolHealth(tool.id)
        self.initializeToolMetrics(tool.id)
        self.circuitBreakers[tool.id] = CircuitBreaker()
        self.recoveryProcedures[tool.id] = RecoveryProcedures()
        allocation = self.resourceAllocations.get(tool.id, ResourceAllocation())
        self._semaphores[tool.id] = asyncio.Semaphore(allocation.concurrencyLimit)

    async def validateToolInstallation(self, tool: Any) -> None:
        return None

    async def executeTool(
        self, tool_id: str, file_paths: List[str], options: Optional[ToolExecutionOptions] = None
    ) -> ToolExecutionResult:
        breaker = self.circuitBreakers.get(tool_id)
        if self._breaker_is_open(breaker):
            raise Exception("Circuit breaker open")

        semaphore = self._semaphores.setdefault(
            tool_id,
            asyncio.Semaphore(
                self.resourceAllocations.get(tool_id, ResourceAllocation()).concurrencyLimit
            ),
        )
        async with semaphore:
            start = time.time()
            try:
                result = await self.executeWithMonitoring(tool_id, file_paths, options)
                self.updateSuccessMetrics(tool_id, result.executionTime, result)
                self._record_breaker_success(tool_id)
                return result
            except Exception as exc:
                self.updateFailureMetrics(tool_id, time.time() - start, exc)
                self._record_breaker_failure(tool_id)
                raise

    async def executeWithMonitoring(
        self, tool_id: str, file_paths: List[str], options: Optional[ToolExecutionOptions] = None
    ) -> ToolExecutionResult:
        return ToolExecutionResult(
            success=True,
            output="[]",
            stderr="",
            executionTime=0.0,
            memoryUsed=0,
            exitCode=0,
            violationsFound=0,
        )

    async def performToolHealthCheck(self, tool_id: str) -> None:
        health = self.healthStatus[tool_id]
        try:
            await self.validateToolInstallation(self.tools[tool_id])
            health.isHealthy = True
            health.healthScore = 100
            health.lastError = None
        except Exception as exc:
            health.isHealthy = False
            health.healthScore = max(0, health.healthScore - 20)
            health.lastError = str(exc)

    async def attemptToolRecovery(self, tool_id: str) -> None:
        procedures = self.recoveryProcedures[tool_id]
        if procedures.resetConfiguration:
            await self.resetToolConfiguration(tool_id)
        if procedures.clearCache:
            await self.clearToolCache(tool_id)
        for step in procedures.customRecoverySteps:
            await self.executeRecoveryStep(tool_id, step)
        self.initializeToolHealth(tool_id)
        self.circuitBreakers[tool_id] = CircuitBreaker()

    async def resetToolConfiguration(self, tool_id: str) -> None:
        return None

    async def clearToolCache(self, tool_id: str) -> None:
        return None

    async def executeRecoveryStep(self, tool_id: str, step: str) -> None:
        return None

    def initializeToolHealth(self, tool_id: str) -> None:
        self.healthStatus[tool_id] = ToolHealth()

    def initializeToolMetrics(self, tool_id: str) -> None:
        self.metrics[tool_id] = ToolMetrics()

    def getToolStatus(self, tool_id: str) -> ToolStatus:
        return ToolStatus(
            tool=self.tools[tool_id],
            health=self.healthStatus[tool_id],
            metrics=self.metrics[tool_id],
            circuitBreaker=self.circuitBreakers[tool_id],
            allocation=self.resourceAllocations.get(tool_id, ResourceAllocation()),
            isRunning=self.getRunningProcessCount(tool_id) > 0,
            queueLength=len(self.executionQueue.get(tool_id, [])),
        )

    def getAllToolStatus(self) -> Dict[str, ToolStatus]:
        return {tool_id: self.getToolStatus(tool_id) for tool_id in self.tools}

    def getToolEnvironment(self, tool: Any) -> ToolEnvironment:
        if tool.id in {"eslint", "tsc"}:
            return self.environments["nodejs"]
        if tool.id in {"flake8", "pylint", "ruff", "mypy", "bandit"}:
            return self.environments["python"]
        return self.environments["system"]

    def prepareExecutionArgs(
        self,
        tool: Any,
        file_paths: List[str],
        config: Optional[ToolConfiguration] = None,
        options: Optional[ToolExecutionOptions] = None,
    ) -> List[str]:
        args = list(getattr(tool, "args", []))
        if config:
            args.extend(config.customArgs)
        if options:
            args.extend(options.additionalArgs)
        args.extend(file_paths)
        return args

    def getRunningProcessCount(self, tool_id: str) -> int:
        prefix = f"{tool_id}_"
        return sum(1 for process_id in self.runningProcesses if process_id.startswith(prefix))

    def countViolationsInOutput(self, tool: Any, output: str) -> int:
        try:
            parsed = json.loads(output)
            if isinstance(parsed, list):
                return sum(len(item.get("messages", [])) if isinstance(item, dict) else 1 for item in parsed)
            if isinstance(parsed, dict):
                return len(parsed.get("messages", parsed.get("violations", [])))
        except Exception:
            pass
        return len([line for line in output.splitlines() if line.strip()])

    def processExecutionQueue(self, tool_id: str) -> None:
        queue = self.executionQueue.get(tool_id, [])
        allocation = self.resourceAllocations.get(tool_id, ResourceAllocation())
        if queue and self.getRunningProcessCount(tool_id) < allocation.concurrencyLimit:
            callback = queue.pop(0)
            callback()

    def updateSuccessMetrics(
        self, tool_id: str, execution_time: float, result: ToolExecutionResult
    ) -> None:
        metrics = self.metrics[tool_id]
        health = self.healthStatus[tool_id]

        metrics.totalExecutions += 1
        metrics.successfulExecutions += 1
        metrics.averageExecutionTime = self._running_average(
            metrics.averageExecutionTime, metrics.totalExecutions, execution_time
        )
        metrics.minExecutionTime = min(metrics.minExecutionTime, execution_time)
        metrics.maxExecutionTime = max(metrics.maxExecutionTime, execution_time)
        metrics.totalViolationsFound += result.violationsFound

        health.successfulExecutions += 1
        health.averageExecutionTime = metrics.averageExecutionTime
        total = health.successfulExecutions + health.failedExecutions
        health.failureRate = health.failedExecutions / total if total else 0.0
        health.isHealthy = True

    def updateFailureMetrics(self, tool_id: str, execution_time: float, error: Exception) -> None:
        metrics = self.metrics[tool_id]
        health = self.healthStatus[tool_id]

        metrics.totalExecutions += 1
        metrics.failedExecutions += 1
        health.failedExecutions += 1
        total = health.successfulExecutions + health.failedExecutions
        health.failureRate = health.failedExecutions / total if total else 0.0
        health.lastError = str(error)
        health.isHealthy = False

    def _running_average(self, previous: float, count: int, value: float) -> float:
        if count <= 1:
            return value
        return ((previous * (count - 1)) + value) / count

    def _breaker_is_open(self, breaker: Any) -> bool:
        if isinstance(breaker, dict):
            return bool(breaker.get("isOpen"))
        return bool(getattr(breaker, "isOpen", False))

    def _record_breaker_success(self, tool_id: str) -> None:
        breaker = self.circuitBreakers.get(tool_id)
        if isinstance(breaker, dict):
            breaker["successCount"] = breaker.get("successCount", 0) + 1
            return
        if breaker:
            breaker.successCount += 1

    def _record_breaker_failure(self, tool_id: str) -> None:
        breaker = self.circuitBreakers.get(tool_id)
        if isinstance(breaker, dict):
            breaker["failureCount"] = breaker.get("failureCount", 0) + 1
            breaker["lastFailureTime"] = time.time()
            if breaker["failureCount"] >= 5:
                breaker["isOpen"] = True
            return
        if breaker:
            breaker.failureCount += 1
            breaker.lastFailureTime = time.time()
            if breaker.failureCount >= 5:
                breaker.isOpen = True
